File: projects/MST_Kruskal_DSU.py
from operator import itemgetter
n, m = [int(i) for i in input().split()]
edges = [[int(i) for i in input().split()] for _ in range(m)]
edges = sorted(edges, key=itemgetter(2), reverse = True)
p = [i for i in range(n)]
r = [0] * n;

def find_dsu( x ):
	return x if x == p[x] else find_dsu( p[x] )

def union_dsu( x, y ):
	tx = find_dsu( x )
	ty = find_dsu( y )
	if r[tx] < r[ty]:
		p[tx] = ty
	elif:
		p[ty] = tx
		if r[tx] == r[ty]:
			r[tx] += 1

result = 0
while(len(edges) > 0):
	edge = edges.pop()
	if find_dsu(edge[0] - 1) != find_dsu(edge[1] - 1):
		result += edge[2]
		union_dsu( edge[0] - 1, edge[1] - 1)
print(result)

File: projects/MST_Kruskal_DSU.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <utility>
using namespace std;

int *p, *r = NULL;

void make_set_dsu( int x ) { p[x] = x; }

int find_dsu( int x ) { return x == p[x] ? x : find_dsu( p[x] ); }

void union_dsu( int x, int y )
{
	x = find_dsu( x );
	y = find_dsu( y );
	if( r[x] < r[y] )
		p[x] = y;
	else
	{
		p[y] = x;
		if( r[x] == r[y] )
			r[x]++;
	}
}

bool cmp(int *a, int *b) { return a[0] < b[0]; }

int main()
{
	int n, m;
	cin >> n >> m;
	int **edges = new int* [m];
	for( int i = 0; i < m; i++ )
	{
		edges[i] = new int[3];
		cin >> edges[i][1] >> edges[i][2] >> edges[i][0];
		edges[i][1]--; edges[i][2]--;
	}

	sort(edges, edges + m, cmp);
	
	p = new int[n];
	r = new int[n];
	for( int i = 0; i < n; i++ ){
		p[i] = i;
	}

	int result = 0;
	for( int i = 0; i < m; i++ )
	{
		int *edge = edges[i];
		if( find_dsu(edge[1]) != find_dsu(edge[2]) )
		{
			result += edge[0];
			union_dsu( edge[1], edge[2] );
		}
	}
	cout << result << endl;
	
}
